# Data loading 

Here we will be using the ```.paraquet``` file we downloaded and do the following:
 - Check metadata and table datatypes of the paraquet file/table
 - Convert the paraquet file to pandas dataframe and check the datatypes. Additionally check the data dictionary to make sure you have the right datatypes in pandas, as pandas will automatically create the table in our database.
 - Generate the DDL CREATE statement from pandas for a sanity check.
 - Create a connection to our database using SQLAlchemy
 - Convert our huge paraquet file into a iterable that has batches of 100,000 rows and load it into our database.

In [ ]:
import pandas as pd 
import pyarrow.parquet as pq
from time import time

In [ ]:
# Read metadata 
pq.read_metadata('yellow_tripdata_2023-09.parquet')

  created_by: parquet-cpp-arrow version 13.0.0
  num_columns: 19
  num_rows: 2846722
  num_row_groups: 3
  format_version: 2.6
  serialized_size: 6357

In [ ]:
# Read file, read the table from file and check schema
file = pq.ParquetFile('yellow_tripdata_2023-09.parquet')
table = file.read()
table.schema

VendorID: int32
tpep_pickup_datetime: timestamp[us]
tpep_dropoff_datetime: timestamp[us]
passenger_count: int64
trip_distance: double
RatecodeID: int64
store_and_fwd_flag: large_string
PULocationID: int32
DOLocationID: int32
payment_type: int64
fare_amount: double
extra: double
mta_tax: double
tip_amount: double
tolls_amount: double
improvement_surcharge: double
total_amount: double
congestion_surcharge: double
Airport_fee: double

In [ ]:
# Convert to pandas and check data 
df = table.to_pandas()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2846722 entries, 0 to 2846721
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[ns]
 2   tpep_dropoff_datetime  datetime64[ns]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee           

We need to first create the connection to our postgres database. We can feed the connection information to generate the CREATE SQL query for the specific server. SQLAlchemy supports a variety of servers.

In [ ]:
# Create an open SQL database connection object or a SQLAlchemy connectable
from sqlalchemy import create_engine

engine = create_engine('postgresql://root:root@localhost:5432/ny_taxi')
engine.connect()

In [ ]:
# Generate CREATE SQL statement from schema for validation
print(pd.io.sql.get_schema(df, name='yellow_taxi_data', con=engine))


CREATE TABLE yellow_taxi_data (
	"VendorID" INTEGER, 
	tpep_pickup_datetime TIMESTAMP WITHOUT TIME ZONE, 
	tpep_dropoff_datetime TIMESTAMP WITHOUT TIME ZONE, 
	passenger_count FLOAT(53), 
	trip_distance FLOAT(53), 
	"RatecodeID" FLOAT(53), 
	store_and_fwd_flag TEXT, 
	"PULocationID" INTEGER, 
	"DOLocationID" INTEGER, 
	payment_type BIGINT, 
	fare_amount FLOAT(53), 
	extra FLOAT(53), 
	mta_tax FLOAT(53), 
	tip_amount FLOAT(53), 
	tolls_amount FLOAT(53), 
	improvement_surcharge FLOAT(53), 
	total_amount FLOAT(53), 
	congestion_surcharge FLOAT(53), 
	"Airport_fee" FLOAT(53)
)




Datatypes for the table looks good! Since we used paraquet file the datasets seem to have been preserved. You may have to convert some datatypes so it is always good to do this check.

## Finally inserting data

There are 2,846,722 rows in our dataset. We are going to use the ```parquet_file.iter_batches()``` function to create batches of 100,000, convert them into pandas and then load it into the postgres database.

In [ ]:
#This part is for testing


# Creating batches of 100,000 for the paraquet file
batches_iter = file.iter_batches(batch_size=100000)
batches_iter

# Take the first batch for testing
df = next(batches_iter).to_pandas()
df

# Creating just the table in postgres
#df.head(0).to_sql(name='ny_taxi_data',con=engine, if_exists='replace')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0             1  2023-09-01 00:15:37   2023-09-01 00:20:21                1   
1             2  2023-09-01 00:18:40   2023-09-01 00:30:28                2   
2             2  2023-09-01 00:35:01   2023-09-01 00:39:04                1   
3             2  2023-09-01 00:45:45   2023-09-01 00:47:37                1   
4             2  2023-09-01 00:01:23   2023-09-01 00:38:05                1   
...         ...                  ...                   ...              ...   
99995         2  2023-09-02 09:55:17   2023-09-02 10:01:45                2   
99996         2  2023-09-02 09:25:34   2023-09-02 09:55:20                3   
99997         2  2023-09-02 09:57:55   2023-09-02 10:04:52                1   
99998         2  2023-09-02 09:35:02   2023-09-02 09:43:28                1   
99999         2  2023-09-02 09:46:09   2023-09-02 10:03:58                1   

       trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0               0.80           1                  N           163   
1               2.34           1                  N           236   
2               1.62           1                  N           162   
3               0.74           1                  N           141   
4               9.85           1                  N           138   
...              ...         ...                ...           ...   
99995           1.48           1                  N           163   
99996          17.49           2                  N           132   
99997           1.73           1                  N           164   
99998           1.32           1                  N           113   
99999           8.79           1                  N           138   

       DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0               230             2          6.5    3.5      0.5        0.00   
1               233             1         14.2    1.0      0.5        2.00   
2               236             1          8.6    1.0      0.5        2.00   
3               229             1          5.1    1.0      0.5        1.00   
4               230             1         45.0    6.0      0.5       17.02   
...             ...           ...          ...    ...      ...         ...   
99995           164             1          9.3    0.0      0.5        2.66   
99996           164             1         70.0    0.0      0.5       24.28   
99997           249             1         10.0    0.0      0.5        2.80   
99998           170             1         10.0    0.0      0.5        4.20   
99999           170             1         35.9    5.0      0.5       10.37   

       tolls_amount  improvement_surcharge  total_amount  \
0              0.00                    1.0         11.50   
1              0.00                    1.0         21.20   
2              0.00                    1.0         15.60   
3              0.00                    1.0         11.10   
4              0.00                    1.0         73.77   
...             ...                    ...           ...   
99995          0.00                    1.0         15.96   
99996          6.94                    1.0        106.97   
99997          0.00                    1.0         16.80   
99998          0.00                    1.0         18.20   
99999          6.94                    1.0         63.96   

       congestion_surcharge  Airport_fee  
0                       2.5         0.00  
1                       2.5         0.00  
2                       2.5         0.00  
3                       2.5         0.00  
4                       2.5         1.75  
...                     ...          ...  
99995                   2.5         0.00  
99996                   2.5         1.75  
99997                   2.5         0.00  
99998                   2.5         0.00  
99999                   2.5         1.75  

[100000 rows x 19 columns]

In [ ]:
# Insert values into the table 
t_start = time()
count = 0
for batch in file.iter_batches(batch_size=100000):
    count+=1
    batch_df = batch.to_pandas()
    print(f'inserting batch {count}...')
    b_start = time()
    
    batch_df.to_sql(name='ny_taxi_data',con=engine, if_exists='append')
    b_end = time()
    print(f'inserted! time taken {b_end-b_start:10.3f} seconds.\n')
    
t_end = time()   
print(f'Completed! Total time taken was {t_end-t_start:10.3f} seconds for {count} batches.')    

inserting batch 1...
inserted! time taken     12.916 seconds.

inserting batch 2...
inserted! time taken     11.782 seconds.

inserting batch 3...
inserted! time taken     11.854 seconds.

inserting batch 4...
inserted! time taken     11.753 seconds.

inserting batch 5...
inserted! time taken     12.034 seconds.

inserting batch 6...
inserted! time taken     11.742 seconds.

inserting batch 7...
inserted! time taken     12.351 seconds.

inserting batch 8...
inserted! time taken     11.052 seconds.

inserting batch 9...
inserted! time taken     12.167 seconds.

inserting batch 10...
inserted! time taken     12.335 seconds.

inserting batch 11...
inserted! time taken     11.375 seconds.

inserting batch 12...
inserted! time taken     10.937 seconds.

inserting batch 13...
inserted! time taken     12.208 seconds.

inserting batch 14...
inserted! time taken     11.542 seconds.

inserting batch 15...
inserted! time taken     11.460 seconds.

inserting batch 16...
inserted! time taken     11

## Extra bit

While trying to do the SQL Refresher, there was a need to add a lookup zones table but the file is in ```.csv``` format. 

Let's code to handle both ```.csv``` and ```.paraquet``` files!

In [ ]:
from time import time
import pandas as pd 
import pyarrow.parquet as pq
from sqlalchemy import create_engine

In [ ]:
url = 'https://d37ci6vzurychx.cloudfront.net/misc/taxi+_zone_lookup.csv'
url = 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-09.parquet'

file_name = url.rsplit('/', 1)[-1].strip()
file_name

'yellow_tripdata_2023-09.parquet'

In [ ]:
if '.csv' in file_name:
    print('yay') 
    df = pd.read_csv(file_name, nrows=10)
    df_iter = pd.read_csv(file_name, iterator=True, chunksize=100000)
elif '.parquet' in file_name:
    print('oh yea')
    file = pq.ParquetFile(file_name)
    df = next(file.iter_batches(batch_size=10)).to_pandas()
    df_iter = file.iter_batches(batch_size=100000)
else: 
    print('Error. Only .csv or .parquet files allowed.')
    sys.exit() 

oh yea


This code is a rough code and seems to be working. The cleaned up version will be in `data-loading-parquet.py` file.